src/libraries/System.Private.CoreLib/src/System/SearchValues/Strings/Helpers/AhoCorasickNode.cs

```diff
@@ -0,0 +1,201 @@
+            if (_firstChildChar >= 0)
+            {
+                queue.Enqueue(((char)_firstChildChar, _firstChildIndex));
+
+                if (_children is Dictionary<char, int> children)
+                {
+                    foreach ((char childChar, int childIndex) in children)
+                    {
+                        queue.Enqueue((childChar, childIndex));
+                    }
+                }
+            }
+        }
+
+        public void OptimizeChildren()
+        {
+            if (_children is Dictionary<char, int> children)
+            {
+                children.Add((char)_firstChildChar, _firstChildIndex);
+
+                float frequency = -2;
+
+                foreach ((char childChar, int childIndex) in children)
+                {
+                    float newFrequency = char.IsAscii(childChar) ? CharacterFrequencyHelper.AsciiFrequency[childChar] : -1;
+
+                    if (newFrequency > frequency)
+                    {
+                        frequency = newFrequency;
+                        _firstChildChar = childChar;
+                        _firstChildIndex = childIndex;
+                    }
+                }
+
+                children.Remove((char)_firstChildChar);
+
+                if (TryCreateJumpTable(children, out int[]? table))
+                {
+                    _children = table;
+                }
+            }
+
+            static bool TryCreateJumpTable(Dictionary<char, int> children, [NotNullWhen(true)] out int[]? table)
+            {
+                // Sacrifice some memory usage in exchange for faster lookup performance
+                const int AcceptableSizeMultiplier = 2;
+
+                Debug.Assert(children.Count > 0);
+
+                int maxValue = -1;
+
+                foreach ((char childChar, _) in children)
+                {
+                    maxValue = Math.Max(maxValue, childChar);
+                }
+
+                int tableSize = TableSizeEstimate(maxValue);
+                int dictionarySize = DictionarySizeEstimate(children.Count);
+
+                if (tableSize > dictionarySize * AcceptableSizeMultiplier)
+                {
+                    // We would have a lot of empty entries. Avoid wasting too much memory.
+                    table = null;
+                    return false;
+                }
+
+                table = new int[maxValue + 1];
+                Array.Fill(table, -1);
+
+                foreach ((char childChar, int childIndex) in children)
+                {
+                    table[childChar] = childIndex;
+                }
+
+                return true;
+
+                static int TableSizeEstimate(int maxValue)
+                {
+                    return 32 + (maxValue * 4);
+                }
+
+                static int DictionarySizeEstimate(int childCount)
+                {
+                    return childCount switch
+                    {
+                        < 4 => 192,
```

Comment:
where do these come from

Reply:
Number of allocated bytes when you create a `Dictionary<char, int>` with a known capacity.
These numbers are current (I think Net 7), but they're just a heuristic so it doesn't matter that much if they aren't matching the exact reality

Reply:
ah - worth a brief comment.